exporter/wavefront/trace.go

```diff
@@ -0,0 +1,178 @@
+package wavefront
+
+import (
+	"bytes"
+	"encoding/hex"
+	"strconv"
+	"sync/atomic"
+
+	wfsender "github.com/wavefronthq/wavefront-sdk-go/senders"
+	"go.opencensus.io/trace"
+)
+
+const (
+	// Tags
+	spanKindKey  = "SpanKind"
+	errTagKey    = "error"
+	errMsgTagKey = "error_msg"
+
+	// Annotations
+	annoMsgKey = "log_msg"
+
+	// Message Events
+	msgIDKey     = "MsgID"
+	msgTypeKey   = "MsgType"
+	msgCmpSzKey  = "MsgCompressedByteSize"
+	msgUcmpSzKey = "MsgUncompressedByteSize"
+)
+
+var (
+	zeroSpanID      = [8]byte(trace.SpanID{})
+	zeroUUID        = []byte("00000000-0000-0000-0000-000000000000")
+	spanKindStrings = [...]string{
+		"unspecified",
+		"server",
+		"client",
+	}
+
+	// Status codes from gRPC.
+	// https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+	statusCodeStrings = [...]string{
+		"OK",
+		"Cancelled",
+		"Unknown",
+		"InvalidArgument",
+		"DeadlineExceeded",
+		"NotFound",
+		"AlreadyExists",
+		"PermissionDenied",
+		"ResourceExhausted",
+		"FailedPrecondition",
+		"Aborted",
+		"OutOfRange",
+		"Unimplemented",
+		"Internal",
+		"Unavailable",
+		"DataLoss",
+		"Unauthenticated",
+	}
+
+	msgEventStrings = [...]string{
+		"unspecified",
+		"sent",
+		"received",
+	}
+)
+
+func (e *Exporter) processSpan(sd *trace.SpanData) {
+	// Span Tags
+	appTags := e.appMap
+	spanTags := make([](wfsender.SpanTag), 0, 3+len(sd.Attributes)+len(appTags))
+	for k, v := range sd.Attributes {
+		spanTags = append(spanTags, wfsender.SpanTag{Key: k, Value: serialize(v)})
+	}
+	for k, v := range appTags {
+		spanTags = append(spanTags, wfsender.SpanTag{Key: k, Value: v})
+	}
+
+	spanKind := sd.SpanKind
+	if spanKind != trace.SpanKindUnspecified {
+		spanTags = append(spanTags, wfsender.SpanTag{Key: spanKindKey,
+			Value: enumString(sd.SpanKind, spanKindStrings[:])})
+	}
+
+	if sd.Status.Code != trace.StatusCodeOK {
+		spanTags = append(spanTags,
+			wfsender.SpanTag{Key: errTagKey, Value: enumString(int(sd.Status.Code), statusCodeStrings[:])},
```

Comment:
In OpenTracing the error tag is either "true" or "false" (okay to omit if false), and there is special handling related to this within Wavefront (at the proxy, UI etc). Let's set the `error` tag to true if there is a error condition, and specify the code under a `error_code` span tag.